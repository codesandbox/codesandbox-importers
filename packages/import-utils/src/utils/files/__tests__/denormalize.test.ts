import denormalize from "../denormalize";

let count = 0;

jest.mock("shortid", () => ({
  generate: () => "" + count++,
}));

describe("denormalize", () => {
  beforeEach(() => {
    count = 0;
  });

  it("can denormalize", () => {
    const paths = {
      "/index.js": { content: "hello", isBinary: false },
      "/world/index.js": { content: "hello2", isBinary: false },
      "/world/hello/index.js": {
        content: "hello3",
        isBinary: false,
        uploadId: "123",
      },
    };

    expect(denormalize(paths)).toMatchSnapshot();
  });

  it("can denormalize with and without leading slash", () => {
    const paths = {
      "index.js": { content: "hello", isBinary: false },
      "world/index.js": { content: "hello2", isBinary: false },
      "world/hello/index.js": { content: "hello3", isBinary: false },
    };

    const slashPaths = {
      "/index.js": { content: "hello", isBinary: false },
      "/world/index.js": { content: "hello2", isBinary: false },
      "/world/hello/index.js": { content: "hello3", isBinary: false },
    };

    const firstDenormalize = denormalize(paths);
    count = 0;

    const secondDenormalize = denormalize(slashPaths);

    expect(firstDenormalize).toEqual(secondDenormalize);
  });

  it("can filter out existing directories", () => {
    const paths = {
      "index.js": { content: "hello", isBinary: false },
      "world/index.js": { content: "hello2", isBinary: false },
      "world/hello/index.js": { content: "hello3", isBinary: false },
      ".github/ISSUE_TEMPLATES/template.md": {
        content: "hello4",
        isBinary: false,
      },
      ".github/workflows/config.yml": { content: "hello5", isBinary: false },
    };

    const existingDirs = [
      {
        directoryShortid: undefined,
        title: "world",
        shortid: "dir1",
      },
      {
        directoryShortid: "dir1",
        title: "hello",
        shortid: "dir2",
      },
      {
        directoryShortid: undefined,
        title: ".github",
        shortid: "dir3",
      },
      {
        directoryShortid: "dir3",
        title: "ISSUE_TEMPLATES",
        shortid: "dir4",
      },
      {
        directoryShortid: "dir3",
        title: "workflows",
        shortid: "dir5",
      },
    ];

    const denormalized = denormalize(paths, existingDirs);

    expect(denormalized).toMatchSnapshot();
    expect(denormalized.directories).toEqual([]);
  });

  it("works with directories starting with a dot", () => {
    const paths = {
      ".github/workflows/test.yml": { content: "test", isBinary: false },
    };

    const denormalized = denormalize(paths, []);

    expect(denormalized).toMatchSnapshot();
    expect(denormalized.directories.length).toBe(2);
  });

  it("can create nested directories", () => {
    const paramFiles = {
      "/src/test/new-file.js": { isBinary: false, content: "" },
    };
    const existingDirs = [
      {
        directoryShortid: null,
        shortid: "rgkK4",
        title: "public",
      },
      {
        directoryShortid: null,
        shortid: "GXOoy",
        title: "src",
      },
    ];
    const denormalized = denormalize(paramFiles, existingDirs);

    expect(denormalized).toMatchSnapshot();
  });

  it("can create only directory", () => {
    const paramFiles: { "/src/test/test2": { type: "directory" } } = {
      "/src/test/test2": { type: "directory" },
    };
    const existingDirs = [
      {
        directoryShortid: null,
        shortid: "rgkK4",
        title: "public",
      },
      {
        directoryShortid: null,
        shortid: "GXOoy",
        title: "src",
      },
    ];
    const denormalized = denormalize(paramFiles, existingDirs);

    expect(denormalized).toMatchSnapshot();
  });
});
