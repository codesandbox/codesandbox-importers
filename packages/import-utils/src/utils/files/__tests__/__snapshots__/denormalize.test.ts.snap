// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`denormalize can create nested directories 1`] = `
{
  "directories": [
    {
      "directoryShortid": "GXOoy",
      "shortid": "0",
      "title": "test",
    },
  ],
  "modules": [
    {
      "code": "",
      "directoryShortid": "0",
      "isBinary": false,
      "sha": undefined,
      "shortid": "1",
      "title": "new-file.js",
      "uploadId": undefined,
    },
  ],
}
`;

exports[`denormalize can create only directory 1`] = `
{
  "directories": [
    {
      "directoryShortid": "GXOoy",
      "shortid": "0",
      "title": "test",
    },
    {
      "directoryShortid": "0",
      "shortid": "1",
      "title": "test2",
    },
  ],
  "modules": [],
}
`;

exports[`denormalize can denormalize 1`] = `
{
  "directories": [
    {
      "directoryShortid": undefined,
      "shortid": "0",
      "title": "world",
    },
    {
      "directoryShortid": "0",
      "shortid": "1",
      "title": "hello",
    },
  ],
  "modules": [
    {
      "code": "hello",
      "directoryShortid": undefined,
      "isBinary": false,
      "sha": undefined,
      "shortid": "2",
      "title": "index.js",
      "uploadId": undefined,
    },
    {
      "code": "hello2",
      "directoryShortid": "0",
      "isBinary": false,
      "sha": undefined,
      "shortid": "3",
      "title": "index.js",
      "uploadId": undefined,
    },
    {
      "code": "hello3",
      "directoryShortid": "1",
      "isBinary": false,
      "sha": undefined,
      "shortid": "4",
      "title": "index.js",
      "uploadId": "123",
    },
  ],
}
`;

exports[`denormalize can filter out existing directories 1`] = `
{
  "directories": [],
  "modules": [
    {
      "code": "hello",
      "directoryShortid": undefined,
      "isBinary": false,
      "sha": undefined,
      "shortid": "0",
      "title": "index.js",
      "uploadId": undefined,
    },
    {
      "code": "hello2",
      "directoryShortid": "dir1",
      "isBinary": false,
      "sha": undefined,
      "shortid": "1",
      "title": "index.js",
      "uploadId": undefined,
    },
    {
      "code": "hello3",
      "directoryShortid": "dir2",
      "isBinary": false,
      "sha": undefined,
      "shortid": "2",
      "title": "index.js",
      "uploadId": undefined,
    },
    {
      "code": "hello4",
      "directoryShortid": "dir4",
      "isBinary": false,
      "sha": undefined,
      "shortid": "3",
      "title": "template.md",
      "uploadId": undefined,
    },
    {
      "code": "hello5",
      "directoryShortid": "dir5",
      "isBinary": false,
      "sha": undefined,
      "shortid": "4",
      "title": "config.yml",
      "uploadId": undefined,
    },
  ],
}
`;

exports[`denormalize works with directories starting with a dot 1`] = `
Object {
  "directories": Array [
    Object {
      "directoryShortid": undefined,
      "shortid": "0",
      "title": ".github",
    },
    Object {
      "directoryShortid": "0",
      "shortid": "1",
      "title": "workflows",
    },
  ],
  "modules": Array [
    Object {
      "code": "test",
      "directoryShortid": "1",
      "isBinary": false,
      "shortid": "2",
      "title": "test.yml",
    },
  ],
}
`;
